ethereum/ethstats/src/main/java/org/hyperledger/besu/ethstats/EthStatsService.java

```diff
@@ -147,22 +147,43 @@ public EthStatsService(
     this.genesisConfigOptions = genesisConfigOptions;
     this.p2PNetwork = p2PNetwork;
     this.blockResultFactory = new BlockResultFactory();
-    this.httpClientOptions = new HttpClientOptions();
-    this.webSocketConnectOptions =
-        new WebSocketConnectOptions()
-            .setURI("/api")
-            .setHost(netstatsUrl.getHost())
-            .setPort(netstatsUrl.getPort())
-            .setSsl(true);
+    this.httpClientOptions = buildHttpClientOptions(netstatsUrl);
+    this.webSocketConnectOptions = buildWebSocketConnectOptions(netstatsUrl);
+  }
+
+  private static HttpClientOptions buildHttpClientOptions(final NetstatsUrl netstatsUrl) {
+    final HttpClientOptions options = new HttpClientOptions();
+
+    if (netstatsUrl.getPemTrust() != null) {
+      options.setPemTrustOptions(
+          new PemTrustOptions().addCertPath(netstatsUrl.getPemTrust().toString()));
+    }
+    return options;
+  }
+
+  private static WebSocketConnectOptions buildWebSocketConnectOptions(
+      final NetstatsUrl netstatsUrl) {
+    return new WebSocketConnectOptions()
+        .setURI("/api")
+        .setSsl(false)
+        .setHost(netstatsUrl.getHost());
+
+    // port will be set up in start method.
+  }
+
+  private static int getWsPort(final NetstatsUrl netstatsUrl, final boolean isSSL) {
+    if (netstatsUrl.getPort() >= 0) {
+      return netstatsUrl.getPort();
+    }
+    return isSSL ? 443 : 80;
   }
 
   /** Start. */
   public void start() {
-
+    flipSSL();
```

Comment:
Why are we flipping it?

Reply:
This doesn't make sense to me either. I thought we would flip after a connection failure. If I understand this correctly if we have SSL specified we will try without SSL, possibly fail then do the connect again flipping to enable SSL again?

Reply:
The naming of the method can be changed. The `reconnect` logic calls `start` method again after n seconds. We want to switch from ssl to non-ssl, and then to ssl on each reconnect attempt.

Reply:
I assumed start is called on startup as well in which case the toggle doesn't make sense to me. Are you saying start is only called first time we reconnect? What if we reconnect more than once?

Reply:
The start is called by the Runner.startExternalServices. So initially we will toggle the SSL. There is also a 10-second delay between each retry. So if we expect to use SSL we try without SSL fail wait for 10s and then connect with SSL. This doesn't seem right to me.

Why do we need to flip-flop between SSL and non-SSL?

Reply:
Thats the behavior of geth. They try to connect to ethstats first with ssl, then with non-ssl. We are replicating the same behavior i.e. first attempt with ssl, next attempt with non-ssl so-on so forth.